components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+    RemoteRecordId, RemoteSuggestion, Result, SuggestApiResult, Suggestion, SuggestionQuery,
+};
+
+const RS_COLLECTION: &str = "quicksuggest";
+
+/// The store is the entry point to the Suggest component. It incrementally
+/// fetches suggestions from the Remote Settings service, stores them in a local
+/// database, and returns them in response to user queries.
+///
+/// Your application should create a single store, and manage it as a singleton.
+/// The store is thread-safe, and supports concurrent fetches and ingests. We
+/// expect that your application will call `query` to show suggestions as the
+/// user types into the address bar, and periodically call `ingest` in the
+/// background to update the database with new suggestions from Remote Settings.
+///
+/// The store keeps track of the state needed to support incremental ingestion,
+/// but doesn't schedule the ingestion work itself, or decide how many
+/// suggestions to ingest at once. This is for two reasons:
+///
+/// 1. The primitives for scheduling background work vary between platforms. You
+///    might use an idle timer on Desktop, `WorkManager` on Android, or
+///    `BGTaskScheduler` on iOS.
+/// 2. Ingestion limits can change, depending on the platform and the needs of
+///    the application. A mobile device on a metered connection might want to
+///    request a small subset of the Suggest data and fetch the rest later,
+///    while a desktop on a fast link might request the entire dataset on first
+///    launch.
+pub struct SuggestStore {
+    path: PathBuf,
+    dbs: OnceCell<SuggestStoreDbs>,
+    settings_client: remote_settings::Client,
+}
+
+/// Limits for an ingestion from Remote Settings.
+pub struct IngestLimits {
+    /// The maximum number of records to request from Remote Settings.
+    /// Each record has about 200 suggestions.
+    pub records: Option<u64>,
+}
+
+impl SuggestStore {
+    /// Creates a Suggest store.
+    pub fn new(
+        path: &str,
+        settings_config: Option<RemoteSettingsConfig>,
+    ) -> SuggestApiResult<Self> {
+        Ok(Self::new_inner(path, settings_config)?)
+    }
+
+    fn new_inner(
+        path: impl AsRef<Path>,
+        settings_config: Option<RemoteSettingsConfig>,
+    ) -> Result<Self> {
+        let settings_client = remote_settings::Client::new(settings_config.unwrap_or_else(|| {
+            RemoteSettingsConfig {
+                server_url: None,
+                bucket_name: None,
+                collection_name: RS_COLLECTION.into(),
+            }
+        }))?;
+        Ok(Self {
+            path: path.as_ref().into(),
+            dbs: OnceCell::new(),
+            settings_client,
+        })
+    }
+
+    /// Returns this provider's database connections, initializing them if
+    /// they're not already open.
+    fn dbs(&self) -> Result<&SuggestStoreDbs> {
+        self.dbs
+            .get_or_try_init(|| SuggestStoreDbs::open(&self.path))
```

Comment:
🗒️ Putting the reader and writer behind a `OnceCell` is a little different than some of our other components, but it lets the constructor avoid doing I/O—so the store is safe to create on the main thread without blocking it.